File: app/src/main/java/com/lucagiorgetti/surprix/utility/dao/DoubleListDao.java
package com.lucagiorgetti.surprix.utility.dao;

import androidx.annotation.NonNull;

import com.google.firebase.database.DataSnapshot;
import com.google.firebase.database.DatabaseError;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.ValueEventListener;
import com.lucagiorgetti.surprix.SurprixApplication;
import com.lucagiorgetti.surprix.listenerInterfaces.CallbackInterface;
import com.lucagiorgetti.surprix.listenerInterfaces.FirebaseListCallback;
import com.lucagiorgetti.surprix.model.Surprise;
import com.lucagiorgetti.surprix.model.User;

import java.util.ArrayList;
import java.util.Objects;

public class DoubleListDao {
    private String username;
    private DatabaseReference userDoubles;
    private DatabaseReference surpriseDoubleOwners;

    public DoubleListDao(String username) {
        this.username = username;
        DatabaseReference reference = SurprixApplication.getInstance().getDatabaseReference();
        this.userDoubles = reference.child("user_doubles").child(username);
        this.surpriseDoubleOwners = reference.child("surprise_doubles");
    }

    public void addDouble(String surpId) {
        userDoubles.child(surpId).setValue(true);
        surpriseDoubleOwners.child(surpId).child(username).setValue(true);
    }

    public void removeDouble(String surpId) {
        userDoubles.child(surpId).setValue(null);
        surpriseDoubleOwners.child(surpId).child(username).setValue(null);
    }

    public void getMissingOwners(String surpId, final FirebaseListCallback<User> listen) {
        listen.onStart();
        final ArrayList<User> owners = new ArrayList<>();
        surpriseDoubleOwners.child(surpId).addListenerForSingleValueEvent(new ValueEventListener() {
            @Override
            public void onDataChange(@NonNull DataSnapshot dataSnapshot) {
                if (dataSnapshot.exists()) {
                    for (DataSnapshot d : dataSnapshot.getChildren()) {
                        UserDao.getUserByUsername(d.getKey(), new CallbackInterface<User>() {
                            @Override
                            public void onStart() {

                            }

                            @Override
                            public void onSuccess(User user) {
                                owners.add(user);
                                listen.onSuccess(owners);
                            }

                            @Override
                            public void onFailure() {

                            }
                        });
                    }
                } else {
                    listen.onSuccess(null);
                }
            }

            @Override
            public void onCancelled(@NonNull DatabaseError databaseError) {

            }
        });
    }

    public void getDoubles(final FirebaseListCallback<Surprise> listen) {
        listen.onStart();
        final ArrayList<Surprise> doubles = new ArrayList<>();

        if (username != null) {
            userDoubles.addListenerForSingleValueEvent(new ValueEventListener() {
                @Override
                public void onDataChange(@NonNull DataSnapshot dataSnapshot) {
                    if (dataSnapshot.exists()) {
                        for (DataSnapshot d : dataSnapshot.getChildren()) {
                            SurpriseDao.getSurpriseById(new CallbackInterface<Surprise>() {
                                @Override
                                public void onStart() {

                                }

                                @Override
                                public void onSuccess(Surprise surprise) {
                                    doubles.add(surprise);
                                    listen.onSuccess(doubles);
                                }

                                @Override
                                public void onFailure() {

                                }
                            }, d.getKey());
                        }
                    } else {
                        listen.onSuccess(null);
                    }
                }

                @Override
                public void onCancelled(@NonNull DatabaseError databaseError) {
                }
            });
        }

    }

    public void clearDoubles() {
        userDoubles.child(username).addListenerForSingleValueEvent(new ValueEventListener() {
            @Override
            public void onDataChange(@NonNull DataSnapshot dataSnapshot) {
                if (dataSnapshot.exists()) {
                    for (DataSnapshot d : dataSnapshot.getChildren()) {
                        surpriseDoubleOwners.child(Objects.requireNonNull(d.getKey())).child(username).setValue(null);
                    }
                }
            }

            @Override
            public void onCancelled(@NonNull DatabaseError error) {

            }
        });

        userDoubles.setValue(null);
    }
}

File: app/src/main/java/com/lucagiorgetti/surprix/utility/dao/DoubleListDao.kt
package com.lucagiorgetti.surprix.utility.dao

import com.google.firebase.database.DataSnapshot
import com.google.firebase.database.DatabaseError
import com.google.firebase.database.DatabaseReference
import com.google.firebase.database.ValueEventListener
import com.lucagiorgetti.surprix.SurprixApplication
import com.lucagiorgetti.surprix.listenerInterfaces.CallbackInterface
import com.lucagiorgetti.surprix.listenerInterfaces.FirebaseCallback
import com.lucagiorgetti.surprix.listenerInterfaces.FirebaseListCallback
import com.lucagiorgetti.surprix.model.Surprise
import com.lucagiorgetti.surprix.model.User
import com.lucagiorgetti.surprix.ui.mainfragments.surpriseList.SurpiseListFirebaseCallback
import java.util.Objects

class DoubleListDao(private val username: String?) {
    private val userDoubles: DatabaseReference
    private val surpriseDoubleOwners: DatabaseReference

    init {
        val reference = SurprixApplication.instance.databaseReference!!
        userDoubles = reference.child("user_doubles").child(username!!)
        surpriseDoubleOwners = reference.child("surprise_doubles")
    }

    fun addDouble(surpId: String?) {
        SurpriseDao.getSurpriseById(object : CallbackInterface<Surprise> {
            override fun onStart() {}
            override fun onSuccess(surprise: Surprise) {
                userDoubles.child(surpId!!).setValue(if (surprise.isSet_effective_code) surprise.code else "ZZZ_$surpId")
                surpriseDoubleOwners.child(surpId).child(username!!).setValue(true)
            }

            override fun onFailure() {}
        }, surpId!!)
    }

    fun removeDouble(surpId: String?) {
        userDoubles.child(surpId!!).setValue(null)
        surpriseDoubleOwners.child(surpId).child(username!!).setValue(null)
    }

    fun getMissingOwners(surpId: String?, listen: FirebaseListCallback<User>) {
        listen.onStart()
        val owners = ArrayList<User>()
        surpriseDoubleOwners.child(surpId!!).addValueEventListener(object : ValueEventListener {
            override fun onDataChange(dataSnapshot: DataSnapshot) {
                if (dataSnapshot.exists()) {
                    for (d in dataSnapshot.children) {
                        UserDao.getUserByUsername(d.key, object : CallbackInterface<User?> {
                            override fun onStart() {}
                            override fun onSuccess(user: User?) {
                                owners.add(user!!)
                                listen.onSuccess(owners)
                            }

                            override fun onFailure() {}
                        })
                    }
                } else {
                    listen.onSuccess(owners)
                }
            }

            override fun onCancelled(databaseError: DatabaseError) {}
        })
    }

    fun getDoubles(listen: SurpiseListFirebaseCallback) {
        listen.onStart()

        userDoubles.orderByValue().addListenerForSingleValueEvent(object : ValueEventListener {
            override fun onDataChange(dataSnapshot: DataSnapshot) {

                if (dataSnapshot.exists()) {
                    listen.onCountDiscovered(dataSnapshot.children.count())
                    for (d in dataSnapshot.children) {
                        SurpriseDao.getSurpriseById(object : CallbackInterface<Surprise> {
                            override fun onStart() {}
                            override fun onSuccess(surprise: Surprise) {
                                listen.onSuccess(surprise)
                            }

                            override fun onFailure() {}
                        }, d.key!!)
                    }
                } else {
                    listen.onFailure()
                }
            }

            override fun onCancelled(databaseError: DatabaseError) {}
        })

    }

    fun clearDoubles() {
        userDoubles.child(username!!).addListenerForSingleValueEvent(object : ValueEventListener {
            override fun onDataChange(dataSnapshot: DataSnapshot) {
                if (dataSnapshot.exists()) {
                    for (d in dataSnapshot.children) {
                        surpriseDoubleOwners.child(Objects.requireNonNull(d.key!!)).child(username).setValue(null)
                    }
                }
            }

            override fun onCancelled(error: DatabaseError) {}
        })
        userDoubles.setValue(null)
    }
}
